Extend X509 cert checks and error reporting in v3_{purp,crld}.c and x509_{set,vfy}.c

DDvO opened the pull request:
As mentioned recently e.g. in https://github.com/openssl/openssl/pull/12357#issuecomment-657407432
I've prepared a PR that adds various checks for not well-formed certs. Here it its.

* add checks for many types of malformednesss to `static int check_chain_extensions(X509_STORE_CTX *ctx)` in `x509_vfc.c`. These are active only if `X509_V_FLAG_X509_STRICT` is set.
* improve error reporting of `X509v3_cache_extensions()` in `v3_purp.c`
* add error reporting to `x509_init_sig_info()` in `x509_set.c`
* improve `static int setup_dp(const X509 *x, DIST_POINT *dp)` and related functions in `v3_purp.c and v3_crld.c`
* add test case for non-conforming certificate from <https://tools.ietf.org/html/rfc8410#section-10.2>

On this occasion I've increased the adherence to the coding guidelines of those functions.
 
- [x] documentation is added or updated
- [x] tests are added or updated

DDvO commented:
@mattcaswell, since you added the option to use a non-default libctx in #11409
it can happen that adding the SHA1 hash to X509 structures fails, for instance because SHA1 has been deprecated or because no suitable implementation is available. So it can happen that `x->sha1_hash` remains NULL without an error being thrown.
Is this on purpose? I've tentatively added a respective TODO to `X509v3_cache_extensions()`:
```
    /* TODO should we really silently skip hash calculation if sha1==NULL ? */
```

mspncp commented:
Nit: typo in commit message:

v3_{purp,crld.c}  ->. v3_{purp,crld}.c

mspncp commented:
Ups, wrong button!

DDvO commented:
> Nit: typo in commit message:
> 
> v3_{purp,crld.c} ->. v3_{purp,crld}.c

Thanks @mspncp for this hint; fixed.
On this occasion I've also updated the title of this PR and extended its description, also in the commit message.
BTW, I would not mind if you review more than the commit header ;-)

DDvO commented:
BTW, after rebasing I found that some CMP CLI test cases started hanging waiting for password input despite the use of `-batch`.
I found that this is due to a wrong dependency in `apps/cmp.c` on engines which meanwhile are deprecated.
So I've added a commit fixing this.

mspncp commented:
> BTW, I would not mind if you review more than the commit header ;-)

Impossible. I was reviewing and posting from my tiny  iPhone, which is the reason why I pressed the wrong button ;-)

DDvO commented:
> > BTW, I would not mind if you review more than the commit header ;-)
> 
> Impossible. I was reviewing and posting from my tiny iPhone, which is the reason why I pressed the wrong button ;-)

Then let's hope you don't hit by mistake the approve or delete button too often ;)

mspncp commented:
What a fantastic proposal for a new GitHub feature:

> the [approve or delete] button
 
(inspired by E. Schrödinger)

DDvO commented:
> What a fantastic proposal for a new GitHub feature:
> the [approve or delete] button

This was no implicit proposal, I was just too lazy to put an 's' behind 'button', 
but anyway a smart idea to quickly deal with PRs ;-)

kaduk commented:
> add checks for many types of malformednesss to `static int check_chain_extensions()`

We will no doubt discover, once this is released, that people are using such malformed certs in production and unhappy about having them suddenly not work.  We should probably think through how to respond to them, in advance, whether that's just to say "sorry; get not-malformed certs" or to have another knob (ugh) to control the behavior, to tell them to set a verify callback, or otherwise.

DDvO commented:
> > add checks for many types of malformednesss to `static int check_chain_extensions()`
> 
> We will no doubt discover, once this is released, that people are using such malformed certs in production and unhappy about having them suddenly not work. We should probably think through how to respond to them, in advance, whether that's just to say "sorry; get not-malformed certs" or to have another knob (ugh) to control the behavior, to tell them to set a verify callback, or otherwise.

Though this would be a valid comment in general 
it looks like you did not read what I wrote immediately thereafter:

> These are active only if `X509_V_FLAG_X509_STRICT` is set.

which also would be obvious from looking at the proposed changes to `check_chain_extensions()`.

kaduk commented:
> > > add checks for many types of malformednesss to `static int check_chain_extensions()`
> > 
> > 
> > We will no doubt discover, once this is released, that people are using such malformed certs in production and unhappy about having them suddenly not work. We should probably think through how to respond to them, in advance, whether that's just to say "sorry; get not-malformed certs" or to have another knob (ugh) to control the behavior, to tell them to set a verify callback, or otherwise.
> 
> Though this would be a valid comment in general
> it looks like you did not read what I wrote immediately thereafter:
> 
> > These are active only if `X509_V_FLAG_X509_STRICT` is set.
> 
> which should also be obvious from looking at the proposed changes to `check_chain_extensions()`.

Sorry for not reading more closely; at times I have a hard time even keeping up with the stream of email from github with updates on issues/PRs, and can't look into the code itself.
That said, `X509_V_FLAG_X509_STRICT` has preexisting behavior, so this is only a partial mitigation for the risk I was worried about.

DDvO commented:
>  That said, `X509_V_FLAG_X509_STRICT` has preexisting behavior, so this is only a partial mitigation for the risk I was worried about.

I don't fully understand what you mean here.
The only thing documented about `X509_V_FLAG_X509_STRICT` is that
it "disables workarounds for some broken certificates", 
and this also covers the new situation with some more such checks.

kaduk commented:
> > That said, `X509_V_FLAG_X509_STRICT` has preexisting behavior, so this is only a partial mitigation for the risk I was worried about.
> 
> I don't fully understand what you mean here.
> The only thing documented about `X509_V_FLAG_X509_STRICT` is that
> it "disables workarounds for some broken certificates",
> and this also covers the new situation with some more such checks.
> 
> Which concern do you still have?

I'm just expressing an opinion that, given the large size of our userbase, it seems plausible to me that someone is running code that currently sets the flag and is using certificates that currently work but will be "broken" by this change.  Given the current documentation that you quote, it seems perfectly reasonable for us to plan to tell them "you have to change your stuff; this is not documented to work for your setup".  I'm just asking that we make that an explicit plan by considering it now, instead of waiting for complaints to come in (and having a risk that if we had known now, instead of when complaints arrive, we might have picked a different option).

kroeckx commented on crypto/x509/x509_set.c:
```diff
@@ -192,8 +192,8 @@ int X509_get_signature_info(X509 *x, int *mdnid, int *pknid, int *secbits,
 
-static void x509_sig_info_init(X509_SIG_INFO *siginf, const X509_ALGOR *alg,
-                               const ASN1_STRING *sig)
+static int x509_sig_info_init(X509_SIG_INFO *siginf, const X509_ALGOR *alg,
```
Since you make it return something, please at least document the return value. A short description would be nice.

DDvO replied:
All right, done.

kroeckx commented:
Before such new checks are added, I would like to the same checks in at least one of the linters (https://github.com/kroeckx/x509lint, https://github.com/zmap/zlint, https://github.com/certlint/certlint).

slontis commented:
Sorry but I needed to add libctx related changes to X509.
You will probably have some nasty merges.

DDvO commented:
>  That said, `X509_V_FLAG_X509_STRICT` has preexisting behavior, so this is only a partial mitigation for the risk I was worried about.

I don't fully understand what you mean here.
The only thing documented about `X509_V_FLAG_X509_STRICT` is that
it "disables workarounds for some broken certificates", 
and this also covers the new situation with some more such checks.


> Sorry but I needed to add libctx related changes to X509.
> You will probably have some nasty merges.

Thanks for your comment.
I like your changes done in PR #12153, and the merge conflicts turned out not that severe.
They also solved a TODO that I referred to above: https://github.com/openssl/openssl/pull/12478#issuecomment-660446462.

DDvO commented:
Oops, my most recent comment contained a large copy&paste mistake, so I deleted it.

> Sorry but I needed to add libctx related changes to X509.
> You will probably have some nasty merges.

Thanks @slontis for your comment.
I like your changes done in PR #12153, and the merge conflicts turned out not that severe.
They also solved a TODO that I referred to above: https://github.com/openssl/openssl/pull/12478#issuecomment-660446462.

DDvO commented:
The new error reporting of `x509_sig_info_init()` revealed that one of the very old test certs is malformed, namely `test/testx509.pem`: 
```
Certificate:
    Data:
        Version: 1 (0x0)
        Serial Number: 24 (0x18)
        Signature Algorithm: md5WithRSAEncryption
        Issuer: C = AU, ST = QLD, CN = SSLeay/rsa test CA
        Validity
            Not Before: Jun 19 23:33:12 1995 GMT
            Not After : Jul 17 23:33:12 1995 GMT
        Subject: C = AU, ST = QLD, CN = SSLeay/rsa test cert
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                RSA Public-Key: (512 bit)
                Modulus:
                    00:aa:db:7a:a9:2e:46:4f:15:71:19:96:16:6b:4f:
                    f8:bb:e2:30:1d:fe:e9:d8:b3:59:6d:c3:c1:a7:df:
                    ce:7c:87:18:01:70:50:9f:c8:4e:fd:17:b5:bb:02:
                    ca:5d:d0:a3:22:86:86:b3:80:cb:74:6f:3c:ae:4c:
                    df:c8:ae:5d:3d
                Exponent: 65537 (0x10001)
    Signature Algorithm: md5
         04:59:ce:c4:70:5f:29:a3:6f:d9:3b:a9:0d:0b:02:bf:20:21:
         fa:0e:86:e0:2d:e9:00:e6:54:8b:af:91:2f:ba:2a:19:9e:26:
         69:e6:51:73:03:0e:3e:63:33:d0:ed:72:89:ce:5f:47:60:83:
         31:01:31:72:a9:28:83:f6:3b:31
```
Note the `Signature Algorithm: md5`, which should contain `md5WithRSAEncryption`.
This made `test_x509` fail. So I've replaced that cert with a valid X.509v1 cert.

DDvO commented:
The new error reporting of `x509_sig_info_init()` also causes `test_rusext` to fail:
```
Unable to load Certificate
4067B521007F0000:error::digital envelope routines:pkey_set_type:unsupported algorithm::0:
4067B521007F0000:error::x509 certificate routines:x509_pubkey_decode:unsupported algorithm::0:
4067B521007F0000:error::x509 certificate routines:x509_sig_info_init:error getting md by nid::0:
```
which I think is correct because `NID_id_GostR3411_94` is known but the implementation of `md_gost94` seems disabled.
So I tried skipping these tests by adding to `test/recipes/25-test_rusext.t`:
```
plan skip_all => "GOST support is disabled in this OpenSSL build"
    if disabled("gost") || disabled("engine");
```
but this was not sufficient for various configurations. 
How to detect/handle this properly?

DDvO commented:
How can one detect using Perl if GOST R 34.11-94 is enabled/available?

t8m commented:
Hmm, you would have to somehow detect whether the gost engine is available. However this error is IMO wrong - the certificate can be properly parsed by openssl so I am not sure the error from the x509 command should be reported. Only when you try to verify the certificate signature it should report such failure.

DDvO commented:
Good point that parsing and analyzing the cert should work independently of the availability of a GOST implementation.

Meanwhile I've found a better solution for GOST R 34.11-94 in `x509_sig_info_init()`, 
namely not to use the theoretical security bits number, which would be `EVP_MD_size(md) * 4` = 256 / 2 = 128, 
but to directly set, as has been done also for SHA1, the currently known number of security bits, 
which is 105 according to https://link.springer.com/chapter/10.1007%2F978-3-540-85174-5_10.

This also makes the tests happy again.

t8m commented:
But the issue that parsing should not abort if used md algo is unavailable is not really fixed?

DDvO commented:
> But the issue that parsing should not abort if used md algo is unavailable is not really fixed?

I've meanwhile fixed this also for other algorithms and added the following remark 
in the resurrected (now internal) documentation of `x509v3_cache_extensions()`:
```
It sets B<X509_SIG_INFO_VALID> in x->flags if x->siginf was filled successfully,
which may not be possible if a referenced algorithm is unknown or not available.
```

DDvO commented:
> Before such new checks are added, I would like to the same checks in at least one of the linters (https://github.com/kroeckx/x509lint, https://github.com/zmap/zlint, https://github.com/certlint/certlint).

@kroeckx, though I agree that something like that would make sense (while I don't know those projects),
I would not like to make this PR on any progress there because doing so would block this PR indefinitely.

DDvO commented:
Rebased the PR, solving merge conflicts.

kroeckx commented:
> > Before such new checks are added, I would like to the same checks in at least one of the linters (https://github.com/kroeckx/x509lint, https://github.com/zmap/zlint, https://github.com/certlint/certlint).
> 
> @kroeckx, though I agree that something like that would make sense (while I don't know those projects),
> I would not like to make this PR on any progress there because doing so would block this PR indefinitely.

So I started implementing some of the checks in x509lint.

kroeckx commented on crypto/x509/v3_purp.c:
```diff
@@ -305,44 +305,62 @@ int X509_supported_extension(X509_EXTENSION *ex)
+        if (dp->distpoint == NULL && sk_GENERAL_NAME_num(dp->CRLissuer) <= 0) {
+            X509err(0, X509_R_INVALID_DISTPOINT);
+            return 0;
+        }
```
I'm unsure why you add this inside the `if (dp->reasons != NULL)`. rfc5280 has this text:
```
   While each of these fields is optional, a DistributionPoint MUST NOT
   consist of only the reasons field; either distributionPoint or
   cRLIssuer MUST be present.
```

Which I guess you translated to that. But I think the important part is: "either distributionPoint or cRLIssuer MUST be present.", it does not depend on the reason being present, you always neither at least one of distributionPoint and cRLIssuer.

DDvO replied:
Ah, you are right! Fixed.

kroeckx commented on crypto/x509/v3_purp.c:
```diff
@@ -305,44 +305,62 @@ int X509_supported_extension(X509_EXTENSION *ex)
+    if (dp->distpoint == NULL && sk_GENERAL_NAME_num(dp->CRLissuer) <= 0) {
+        X509err(0, X509_R_INVALID_DISTPOINT);
+        return 0;
+        }
```
The } seems to have a wrong indentation.

DDvO replied:
Oops, fixed.

kroeckx commented on crypto/x509/v3_purp.c:
```diff
@@ -305,44 +305,62 @@ int X509_supported_extension(X509_EXTENSION *ex)
     int i;
 
-    if (dp->reasons) {
+    if (dp->distpoint == NULL && sk_GENERAL_NAME_num(dp->CRLissuer) <= 0) {
```
I'm currently not sure if == 0 is a problem or not. Below you seem to say that <= 1 is acceptable, not == 1. You can change it to `dp->CRLissuer == NULL`

kroeckx replied:
I currently can't see how == 0 can make any sense, so I guess it's fine.

DDvO replied:
> I'm currently not sure if == 0 is a problem or not.

It is - according to the 2nd paragraph of section 4.2.1.13., if distributionPoint is missing the cRLissuer must contain at least one entry:

> If the certificate issuer is not the CRL
>    issuer, then the cRLIssuer field MUST be present and contain the Name
>    of the CRL issuer.  If the certificate issuer is also the CRL issuer,
>    then conforming CAs MUST omit the cRLIssuer field and MUST include
>    the distributionPoint field


> Below you seem to say that <= 1 is acceptable, not == 1

That refers to a different requirement further down in the same section:

> The
>    DistributionPointName MUST NOT use the nameRelativeToCRLIssuer
>    alternative when cRLIssuer contains more than one distinguished name.


> You can change it to `dp->CRLissuer == NULL`

Not really, because as just mentioned if dp->distpoint == NULL then at least one cRLissuer element must be present.

kroeckx replied:
So I've added some new checks to x509 that should check the same:
https://github.com/kroeckx/x509lint/blob/2e2613e5953d37fb6ffb27cff94cb55167c6c533/checks.c#L1164-L1177
https://github.com/kroeckx/x509lint/blob/2e2613e5953d37fb6ffb27cff94cb55167c6c533/checks.c#L1206-L1209

DDvO replied:
> So I've added some new checks to x509 that should check the same:

Nice.

> https://github.com/kroeckx/x509lint/blob/2e2613e5953d37fb6ffb27cff94cb55167c6c533/checks.c#L1164-L1177

The 
```
			if (dp->distpoint == NULL && dp->CRLissuer == NULL)
			{
				SetError(ERR_INVALID_CRL_DIST_POINT);
			}
			if (dp->CRLissuer != NULL)
			{
				if (sk_GENERAL_NAME_num(dp->CRLissuer) == 0)
				{
					SetError(ERR_INVALID_CRL_DIST_POINT);
				}
```
does not cover the case that CRLissuer is present but empty. So should be as I use in this PR:
```
if (dp->distpoint == NULL &&  sk_GENERAL_NAME_num(dp->CRLissuer) <= 0)
			{
				SetError(ERR_INVALID_CRL_DIST_POINT);
			}
```

> https://github.com/kroeckx/x509lint/blob/2e2613e5953d37fb6ffb27cff94cb55167c6c533/checks.c#L1206-L1209

The `if (dp->CRLissuer != NULL && sk_GENERAL_NAME_num(dp->CRLissuer) != 1)`
can be simplified to `if (sk_GENERAL_NAME_num(dp->CRLissuer) <= 1)`.

kroeckx replied:
> does not cover the case that CRLissuer is present but empty.

I'm not sure what you're saying, since that seems to be exactly what's checked.

DDvO replied:
According to RFC 5280 and
```
        ASN1_IMP_SEQUENCE_OF_OPT(DIST_POINT, CRLissuer, GENERAL_NAME, 2)
```
the CRLissuer list is optional, so can be absent
(`dp->CRLissuer == NULL` or equivalently `sk_GENERAL_NAME_num(dp->CRLissuer) == -1`), 
and at least according to the above OpenSSL declaration it can be present but empty 
(`sk_GENERAL_NAME_num(dp->CRLissuer) == 0`).
Of course, but cases mean that no CRLissuer is included.

I caught both cases by `sk_GENERAL_NAME_num(dp->CRLissuer) <= 0`.

DDvO replied:
Ah yes, together with the guard `if (dp->CRLissuer != NULL)`
your code does cover the case that the CRLissuer field is present but empty.

kroeckx commented:
So I think I've modified x509lint to cover all the new errors, and
some more. I've started running this on my certificates database.
After having processed 1M certificates, I see the following errors:
err_akid_missing: 422
err_not_all_crl_reasons: 6

For all the ones I've looked at, they seem to be test certificates
from an untrusted roots.

kroeckx commented:
I'm now at 30M certificates tested, and of what is in the CT logs
I don't have any concerns currently.

But I have an /etc/ssl/certs/dovecot.pem (from 2014) that gets:
E: No key usage
E: CA certificate with non-critical Basic Constraints
I: Checking as root CA certificate

I have no idea if dovecot ever fixed it's generation of
certificates, but I know that things like outlook and thunderbird
accept such certificates. But it seems you didn't add such checks,
you marked the non-critical Basic Constraints as a TODO, and I don't
see you adding a check for no key usage, only an empty one.

So I think I need to wait until it gets to certs that it didn't
get from CT to see if something will break.

kroeckx commented:
So looking at other certs seen in the wild, I see various
certificates that have the combination of no key usage, no basic
constraints, no skid, where at least the skid check is now part of
the strict check.

DDvO commented:
> I'm now at 30M certificates tested, and of what is in the CT logs I don't have any concerns currently.

Wow, what a large amount of test certs!

> But I have an /etc/ssl/certs/dovecot.pem (from 2014) that gets: 
> E: No key usage 

RFC 5280 requires the key usage extension only for (typical) CA certs:

> Conforming CAs MUST include this extension in certificates that
   contain public keys that are used to validate digital signatures on
   other public key certificates or CRLs.

So I suggest re-wording the error message to
E: No key usage for CA certificate
or
E: CA certificate without key usage

> E: CA certificate with non-critical Basic Constraints

Hmm, this error, as well as the one given above, should only be given for 
"CA certificates that contain public keys used to validate digital signatures on certificates".
How can a tool know about this (unless the cert in question is being used during cert chain validation)?
Well, on the other hand I can hardly imagine a CA cert not used for this purpose.
Still one might consider downgrading those errors to warnings
or adding a verification option to explicitly indicate that CA certs are supposed to be used for cert or CRL verification
or to make that the default but add an option to disable this assumption and thus those types of errors.

> I: Checking as root CA certificate I have no idea if dovecot ever fixed it's generation of certificates, but I know that things like Outlook and Thunderbird accept such certificates.

When using such certs for cert chain validation, such certs should not be acceptable,
but maybe all sorts of checks are disabled anyway on certs explicitly trusted?

> But it seems you didn't add such checks, you marked the non-critical Basic Constraints as a TODO,

All right, just solved this TODO, as well as tree further ones: AKID and SKID must not be marked critical,
and if the subject is empty the SAN extension must be marked critical.

> and I don't see you adding a check for no key usage, only an empty one.

Yeah, I had thought at first that the non-emptyness of the key usage for CA certs is checked indirectly already,
but this was not really the case; done now.

DDvO commented:
> So looking at other certs seen in the wild, I see various certificates that have the combination of no key usage, no basic constraints, no skid, where at least the skid check is now part of the strict check.

For EE certs those types of extensions are not required.
I assume you mean CA certs with those defects (where for CA certs the problem cannot be the absence of basic constraints but their missing criticality marker).

kroeckx commented:
x509lint currently doesn't really work properly on self-signed
non-CA certificate, but as we have already covered in a different
PR, most of rfc5280 doesn't really apply to such certs. But you're
now adding new checks based on rfc5280, and are applying it to certs
that don't need to conform to it.

As far as I know, placing just a self-signed non-CA in the trust
store should just work to verify the chain, but I think this PR adds
more checks and that some of them will cause breakage. Maybe the
whole strict check part should depend on num > 0.

kroeckx commented:
> > I'm now at 30M certificates tested, and of what is in the CT logs I don't have any concerns currently.
> 
> Wow, what a large amount of test certs!

I actually have more than 2G certificates.

> > But I have an /etc/ssl/certs/dovecot.pem (from 2014) that gets: 
> > E: No key usage 

That dovecot cert is really weird, it only has 3 extentions:
- SKID
- AKID
- CA:TRUE

But there is no reason for it to be a CA.

kroeckx commented:
I think #12140 get fixed by this PR.

It would be nice that we actually have tests for things we want to reject.

kroeckx commented:
#12141 is probably also fixed.

DDvO commented:
> I think #12140 get fixed by this PR.
> 
> It would be nice that we actually have tests for things we want to reject.

> #12141 is probably also fixed.

Not really. We need to distinguish general certificate well-formedness, which is the topic of this PR,
and whether the actual use of keys contained in certs is allowed by their key usage entries.
See my new bug report #12721 generalizing the two bug reports referred to above.

DDvO commented:
> x509lint currently doesn't really work properly on self-signed non-CA certificate,

Apparently this holds for various crypto software, including the NSS lib used by Mozilla Thunderbird, Firefox, etc.
For instance, I took part in discussions on this topic on Bugzilla and even offered a pragmatic patch to them.
https://bugzilla.mozilla.org/show_bug.cgi?id=202329
https://bugzilla.mozilla.org/show_bug.cgi?id=1523130
but those guys do not care much about it, and I doubt their experts really understand the topic and its implications.

> but as we have already covered in a different PR, most of rfc5280 doesn't really apply to such certs.

Yes, and the same hold more generally for self-issued non-CA certs.

> But you're now adding new checks based on rfc5280, and are applying it to certs that don't need to conform to it.

Very good point - I was not aware of that when preparing this PR where I did not think of those types of certs.


> As far as I know, placing just a self-signed non-CA in the trust store should just work to verify the chain, but I think this PR adds more checks and that some of them will cause breakage.

Some of the new checks only pertain to CA certs, but indeed there are others that affect also self-issued EE certs.

> Maybe the whole strict check part should depend on num > 0.

That would be too much of an exception, so I've refined this idea.
I've just added a condition to make sure that the strict cert checks are not done for self-issued non-CA certs
and added a comment that this is because RFC 5280 does not apply to them according RFC 6818 section 2.

kroeckx commented:
So I've made some other changes to x509lint and restarted going
over the database. I've put some current stats at:
https://www.roeckx.be/x509lint_stats.txt

DDvO commented:
> Before such new checks are added, I would like to the same checks in at least one of the linters (https://github.com/kroeckx/x509lint, https://github.com/zmap/zlint, https://github.com/certlint/certlint).

@kroeckx, can this PR be approved now?
Would be nice if this gets merged before beta1.

kroeckx commented on crypto/x509/v3_purp.c:
```diff
@@ -398,30 +417,34 @@ int x509v3_cache_extensions(X509 *x)
     }
     ERR_set_mark();
 
+    /* Cache the SHA1 digest of the cert */
```
I'm not convinced that X509_digest() will cache the result if it's not cached yet.

DDvO replied:
> I'm not convinced that X509_digest() will cache the result if it's not cached yet.

Just have a look at the code: it calls `X509_digest()` only in a situation where `x->ex_flags & EXFLAG_SET` is not yet set, and in this case `X509_digest()` does call `asn1_item_digest_with_libctx()`, which will set `x->sha1_hash`.

kroeckx replied:
So looking closer, when it is already cached, we do a memcpy() with dest and src the same, which is undefined behaviour, but probably works in practice.

DDvO replied:
The `memcpy()` call is done only when the hash value was already in place, which does not happen when called from `x509v3_cache_extensions()` because this is done only in case `x->ex_flags & EXFLAG_SET` is not true.

kroeckx commented on crypto/x509/v3_purp.c:
```diff
@@ -398,30 +417,34 @@ int x509v3_cache_extensions(X509 *x)
+        x->ex_flags |= EXFLAG_INVALID;
+    /*
+     * Note that the cert is marked invalid also on internal malloc failure
+     * or on failure of EVP_MD_fetch(), potentially called by X509_digest().
```
Is this some sort of TODO? In any case, I think it makes more sense to put the comment above the code.

DDvO replied:
It can be seen as a TODO because it would be better to improve the situation described.
I've moved the comment above the ` x->ex_flags |= EXFLAG_INVALID;`.

kroeckx commented on crypto/x509/v3_purp.c:
```diff
@@ -398,30 +417,34 @@ int x509v3_cache_extensions(X509 *x)
         if (bs->pathlen != NULL) {
+            /*
+             * the error case !bs->ca is checked by check_chain_extensions()
+             * in case ctx->param->flags & X509_V_FLAG_X509_STRICT
```
I'm not sure why you removed the check here and made it only part of the strict check, making OpenSSL tolerate more invalid certificates.

I'm not sure why OpenSSL allowed a non-CA certificate to have a pathlen of 0, since that doesn't make sense, the last CA in the chain has a pathlen of 0.

DDvO replied:
Indeed the original form of the check was wrong - I guess whoever implemented this did not understand the semantics of pathlen == 0.

I moved the improved check to `check_chain_extensions()` for consistency with the other checks.
As far as I've learned doing this PR, `x509v3_cache_extensions()` basically does only those checks required for a cert to be usable in the crypto lib, while the non-essential checks are done in  `check_chain_extensions()`, where many of them are only done if strict checking is requested.

kroeckx replied:
You could argue that a function that does caching shouldn't even try to mark the certificate as invalid.

DDvO replied:
I agree, yet on the other hand maybe the caller expects the SHA1 hash to be present when `x509v3_cache_extensions()` returns success, i.e., does not mark the cert is invalid. Yet this and many further such details were not documented, while this PR at least adds some basic documentation in `doc/internal/man3/x509v3_cache_extensions.pod`.

kroeckx commented on crypto/x509/v3_purp.c:
```diff
@@ -436,17 +459,17 @@ int x509v3_cache_extensions(X509 *x)
-        if (pci->pcPathLengthConstraint) {
+        if (pci->pcPathLengthConstraint != NULL)
             x->ex_pcpathlen = ASN1_INTEGER_get(pci->pcPathLengthConstraint);
-        } else
```
The `{` doesn't seem needed here.

DDvO replied:
Yes, and that's why I removed it, also for adherence to the coding guidelines.

kroeckx replied:
Clearly I was confused.

kroeckx commented on crypto/x509/v3_purp.c:
```diff
@@ -550,19 +587,37 @@ int x509v3_cache_extensions(X509 *x)
 
-    x509_init_sig_info(x);
+    /* Set x->siginf, ignoring errors due to unsupported algos */
+    (void)x509_init_sig_info(x);
```
I'm don't understand why you make x509_init_sig_info() return an error, and then have the only caller ignore the error. I'm also unsure why we need x509_init_sig_info(), since it seems to just call x509_sig_info_init(), which has the same words in the name but in a different order. And x509_sig_info_init() also changed from returning void to an int, and also only has 1 caller.

DDvO replied:
> I'm don't understand why you make x509_init_sig_info() return an error, and then have the only caller ignore the error.

Originally I wanted to make use of the error, but it turned out better not to do.
Yet for some other/future use of `x509_init_sig_info()` it may be useful to get an error returned.

>  I'm also unsure why we need x509_init_sig_info(), since it seems to just call x509_sig_info_init(), which has the same words in the name but in a different order. And x509_sig_info_init() also changed from returning void to an int, and also only has 1 caller.

I did not design this, but apparently the idea was that `x509_sig_info_init()` is more general such that it could be used also for handling other structures. For instance, in `crypto/x509/` there are other such cases where part of X509 and X509_CRL structures are treated uniformly.

DDvO commented:
Thanks @kroeckx for your interesting new comments.
I think I've meanwhile answered all of them.

kroeckx commented on crypto/x509/x509_vfy.c:
```diff
@@ -519,12 +520,72 @@ static int check_chain_extensions(X509_STORE_CTX *ctx)
+         */
+        if ((ctx->param->flags & X509_V_FLAG_X509_STRICT) != 0
+            && !(i == 0 && (x->ex_flags & EXFLAG_CA) == 0
+                 && (x->ex_flags & EXFLAG_SI) != 0)) {
```
I don't think this is the check we want to do. I think we need to look at num > 1 instead.

DDvO replied:
Using
```
     && !(i == 0 && (x->ex_flags & EXFLAG_CA) == 0
                 && (x->ex_flags & EXFLAG_SI) != 0))
```
I tried to follow https://tools.ietf.org/html/rfc6818#section-2 very closely, which says:

> use of self-issued certificates and self-signed certificates
> issued by entities other than CAs are outside the scope of this specification

Using instead `num > 1` is rather indirect in my view, but seems to have pretty much the same effect in practice.
So I've changed the condition this way.

kroeckx commented on crypto/x509/x509_vfy.c:
```diff
@@ -519,12 +520,72 @@ static int check_chain_extensions(X509_STORE_CTX *ctx)
+            /* Check Key Usage according to RFC 5280 section 4.2.1.3 */
+            if ((x->ex_flags & EXFLAG_CA) != 0) {
+                if ((x->ex_flags & EXFLAG_KUSAGE) == 0)
+                    ctx->error = X509_V_ERR_CA_CERT_MISSING_KEY_USAGE;
```
Should we also check for KU_KEY_CERT_SIGN?

DDvO replied:
This is not required for general cert conformance, and it depends on the actual usage of a CA cert if keyCertSign and/or cRLSign needs to be set.

kroeckx replied:
I'm not sure how a CA cert can sign a CRL without also being able to sign a cert, but I have to admit that I don't fully understand how CRL signing certs work currently.

DDvO replied:
See RFC 5280 seciion 5.1.1.3: 
> CAs that are also CRL issuers MAY use one private key to digitally
   sign certificates and CRLs, or MAY use separate private keys to
   digitally sign certificates and CRLs.  When separate private keys are
   employed, each of the public keys associated with these private keys
   is placed in a separate certificate, one with the keyCertSign bit set
   in the key usage extension, and one with the cRLSign bit set in the
   key usage extension (Section 4.2.1.3).

kroeckx replied:
So check for KU_KEY_CERT_SIGN|KU_CRL_SIGN?

kroeckx replied:
I guess the RFC doesn't really say anything about that.

kroeckx replied:
But I still expect KU_KEY_CERT_SIGN for i > 0

DDvO replied:
> I guess the RFC doesn't really say anything about that.

As I wrote above, this is not part of the general (static) conformance requirements in section 4, which are meant to be implemented by `check_chain_extensions()`.
The RFC does require checking keyCertSign and cRLSign elsewhere, namely in section 6 which focuses on the dynamic cert chain checks when the actual use of certs is verified.

DDvO replied:
The check for KU_KEY_CERT_SIGN is already implemented in `x509_signing_allowed()`, called by `internal_verify()`.
Both `check_chain_extensions()` and `internal_verify()` are called by `verify_chain()`,
so I see no need to check KU_KEY_CERT_SIGN also in `check_chain_extensions()`.

kroeckx commented on crypto/x509/x509_vfy.c:
```diff
@@ -519,12 +520,72 @@ static int check_chain_extensions(X509_STORE_CTX *ctx)
+            if (x->akid != NULL && (x->ex_flags & EXFLAG_AKID_CRITICAL) != 0)
+                ctx->error = X509_V_ERR_AUTHORITY_KEY_IDENTIFIER_CRITICAL;
+            if (x->akid != NULL && (x->ex_flags & EXFLAG_SKID_CRITICAL) != 0)
+                ctx->error = X509_V_ERR_SUBJECT_KEY_IDENTIFIER_CRITICAL;
```
This should check x->skid, akid was the previous if.

DDvO replied:
Oops, fixed.

kroeckx commented on crypto/x509/x509_vfy.c:
```diff
@@ -519,12 +520,72 @@ static int check_chain_extensions(X509_STORE_CTX *ctx)
+                                 * taken as "generated by conforming CAs"
+                                 */
+                        && (x->akid == NULL || x->akid->keyid == NULL))
+                    ctx->error = X509_V_ERR_MISSING_AUTHORITY_KEY_IDENTIFIER;
```
The lasts cert in the chain might not be the root CA but an intermediate CA that's trusted. I see no reason to not check it for trusted intermediate CAs.

I'm not sure the ecosystem is ready for this check, I have at least 5M non-self signed certificates that don't have an AKID.

kroeckx replied:
Checking various of those certificates, I'm not concerned anymore currently.

kroeckx commented on crypto/x509/x509_vfy.c:
```diff
@@ -519,12 +520,72 @@ static int check_chain_extensions(X509_STORE_CTX *ctx)
+                ctx->error = X509_V_ERR_AUTHORITY_KEY_IDENTIFIER_CRITICAL;
+            if (x->akid != NULL && (x->ex_flags & EXFLAG_SKID_CRITICAL) != 0)
+                ctx->error = X509_V_ERR_SUBJECT_KEY_IDENTIFIER_CRITICAL;
+            if (X509_get_version(x) >= 2) { /* at least X.509v3 */
```
I think we should just generate an error in case it's not at least a v3, I don't see a reason to only do that check for v3.

DDvO replied:
As far as I understand, v1 or v2 certs cannot have extensions such as SKID and AKID,
so without that guard we would always report non-conformance for certs not having at least version 3.

kroeckx replied:
You're right that in 4.1.2.1 it actually allows version 1 and 2 certificates in special cases. But for a CA we require that it has a key usage extension above. The AKID is also a MUST for non-root CAs in the RFC.

DDvO replied:
> You're right that in 4.1.2.1 it actually allows version 1 and 2 certificates in special cases.

Yes.

> But for a CA we require that it has a key usage extension above.

We check for existence of key usage only for certs with EXFLAG_CA, which can only be set if a basic key usage extension is present, which can happen only for certs with version at least 3. So here we do not need the explicit guard `X509_get_version(x) >= 2`.

> The AKID is also a MUST for non-root CAs in the RFC.

That's right, and this test is included as part of this PR (or course, only for at least v3 as discussed).

kroeckx replied:
I was under the impression that EXFLAG_CA was set in all cases X509_check_ca said it was a CA, but clearly it's only when the basic constrains section exists.

Do we have an places where we enforce that extensions are only allowed for v3 certs?

DDvO replied:
Good point - such a check was missing; I've just added it.

DDvO commented:
Thanks @kroeckx for your further comments, which motivated two further small changes.
For the three remaining new comments I don't think a change is due.
Can you approve this PR now, such that it can be merged before feature freeze and will make it into 3.0?

DDvO commented:
Can this be approved now?

DDvO commented:
@kroeckx, would be really good if we can finalize this PR now after nearly 8 weeks.

The hitherto checks on certificates have been far from being well-structured, systematic, and complete,
and this PR has certainly improved the situation but it does not need to be perfect.
Any further completions can still be done in subsequent PRs.

t8m commented:
I would approve but @kroeckx has much better insight into the changes so his approval would be more valuable.

kroeckx commented:
It's propbably ok, but I need to find some time to do the final
review. I'll try to find some time this evening.

t8m reviewed (approved):
Approving anyway

t8m commented:
But please wait with merging for @kroeckx final review.

kroeckx commented on include/openssl/x509v3.h:
```diff
@@ -375,8 +374,12 @@ struct ISSUING_DIST_POINT_st {
+# define EXFLAG_BCONS_CRITICAL   0x10000
+# define EXFLAG_AKID_CRITICAL    0x20000
+# define EXFLAG_SKID_CRITICAL    0x40000
+# define EXFLAG_SAN_CRITICAL     0x80000
```
Is it useful to put this in a public header? As far as I know, you can't get to those flags anymore since 1.1.0.

kroeckx replied:
But putting them in an other header will probably lead to confusion.

DDvO replied:
Well, I put them there simply because also all other `#define EXFLAG_*` have been placed there, too.
Most of them are documented and may be of good external use, 
while others, including these four but also EXFLAG_NSCERT and EXFLAG_SET, 
but also the documented flags EXFLAG_CRITICAL and EXFLAG_INVALID_POLICY, are likely only used internally.
This could be cleaned up in a later PR.

DDvO commented:
Thanks a lot again @kroeckx for your thorough review and the interesting discussion!
Very pleased that this PR is finally approved, still in time for he beta1 release :tada:

DDvO commented:
Merged :smile:
